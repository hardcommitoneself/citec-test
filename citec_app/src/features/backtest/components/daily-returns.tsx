// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { Ts } from 'features/utils/interfaces/backtest-api-resp'
import moment from 'moment'
import ReactApexChart from 'react-apexcharts'

interface DailyReturnsProps {
   ts_backtest: Ts
}

export const DailyReturns = ({ ts_backtest }: DailyReturnsProps) => {
   const { benchmark, portfolio, date } = ts_backtest

   const portfolioConverted = portfolio.map((num) => Number((num * 100).toFixed(3)))
   const benchmarkConverted = benchmark.map((num) => Number((num * 100).toFixed(3)))
   // Y AXIS

   const series = [
      {
         name: 'Portfolio',
         data: portfolioConverted,
      },
      {
         name: 'Benchmark',
         data: benchmarkConverted,
      },
   ]

   // const type = "line" as const;

   const options: ApexCharts.ApexOptions = {
      chart: {
         type: 'line' as const,
         height: 350,
         background: '#f6f7fb',
      },
      dataLabels: {
         enabled: false,
      },
      labels: date,
      stroke: {
         curve: 'straight' as const,
         width: 1.5,
      },
      grid: {
         row: {
            colors: ['#f3f3f3', 'transparent'], // takes an array which will be repeated on columns
            opacity: 0.5,
         },
      },
      xaxis: {
         tickAmount: 11,
         categories: [],
         labels: {
            rotate: 0,
            formatter(value) {
               const formatedDate = moment(value, 'DD-MM-YYYY').format('MM-YYYY')

               return formatedDate
            },
         },
      },
      yaxis: {
         tickAmount: undefined, // apexcharts.com/docs/options/yaxis/
         floating: false,
         decimalsInFloat: 0,
      },
   }

   return (
      <ReactApexChart options={options} series={series} type="line" height={350} />
   )
}
