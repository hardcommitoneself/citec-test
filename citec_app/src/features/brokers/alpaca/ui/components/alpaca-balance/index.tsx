// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { useDispatch } from 'react-redux'
import { LinkAccount } from 'features/brokers/link-account'
import MenuItem from '@mui/material/MenuItem'
import FormControl from '@mui/material/FormControl'
import Select, { SelectChangeEvent } from '@mui/material/Select'
import { setAccountTypeSelect } from 'store/slices/brokers'
import { useAppSelector } from 'store/hooks'
import { HistoryPortfolio } from './alpaca-history-portfolio'

export const AlpacaBalance = () => {
   const visibility = process.env.REACT_APP_ALPACA_VISIBILITY

   const dispatch = useDispatch()

   const brokers = useAppSelector((state) => state.brokers)
   const typesAccount = brokers.accountType

   const handleChange = (event: SelectChangeEvent) => {
      const typeSelected = event.target.value as string
      dispatch(setAccountTypeSelect({ accountTypeSelect: typeSelected }))
   }

   return (
      <>
         {visibility == 'true' && (
            <div className="mx-10 mt-2 bg-white p-4 rounded-2xl relative balance">
               <div className="brokerTitle">
                  <div className="actionBroker">
                     <div className="selectAccount">
                        <span className="text-sm">Account</span>
                        <FormControl
                           variant="standard"
                           sx={{ m: 1, width: 200, pt: 0 }}
                           size="small"
                        >
                           <Select
                              value={brokers.accountTypeSelect}
                              label="type-acount"
                              onChange={handleChange}
                              inputProps={{ 'aria-label': 'Without label' }}
                              className="selectAccountType"
                           >
                              {typesAccount.map((typeAc) => (
                                 <MenuItem value={typeAc} key={typeAc}>
                                    {typeAc == 'PAPER'
                                       ? 'Paper Trading'
                                       : 'Live Trading'}
                                 </MenuItem>
                              ))}
                           </Select>
                        </FormControl>
                     </div>

                     <LinkAccount />
                  </div>
               </div>
               <HistoryPortfolio />
            </div>
         )}
      </>
   )
}
