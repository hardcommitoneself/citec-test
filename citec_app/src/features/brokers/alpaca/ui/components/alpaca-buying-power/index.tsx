// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { useState, useEffect } from 'react'
import { Button } from '@mui/material'
import { formatToMoney } from 'features/utils/format-to-money'
import { AlpacaApi } from 'features/brokers/alpaca/utils/hooks/api/alpaca-api'
import { useDispatch } from 'react-redux'
import { resetBrokers } from 'store/slices/brokers'
import { useAppSelector } from 'store/hooks'

export const AlpacaBuyingPower = () => {
   const dispatch = useDispatch()

   const visibility = process.env.REACT_APP_ALPACA_VISIBILITY
   const [buyingPower, setBuyingPower] = useState(0)

   const brokers = useAppSelector((state) => state.brokers)

   useEffect(() => {
      getBuyinPower(brokers.accountTypeSelect)
      const interval = setInterval(() => {
         getBuyinPower(brokers.accountTypeSelect)
      }, 5000)

      return function cleanup() {
         clearInterval(interval)
      }
   }, [brokers.accountTypeSelect])

   const getBuyinPower = (type: string) => {
      AlpacaApi({
         type: type,
         url: 'account',
         method: 'GET',
      })
         .then(function (response) {
            setBuyingPower(response.data.buying_power)
         })
         .catch(function (error) {
            if (error.response) {
               if (error.response.status == '403') {
                  console.log(error.response.status)
                  dispatch(resetBrokers())
                  localStorage.removeItem('a1c_t')
               } else {
                  console.log(error.response.data)
                  console.log(error.response.status)
                  console.log(error.response.headers)
               }
            }
         })
   }

   return (
      <>
         {visibility == 'true' && (
            <div className="mx-10 mt-8 bg-white p-4 rounded-2xl relative balance">
               <div className="flex buying-power">
                  <h3 className="font-semibold text-[26px] text-[#2d405a]">
                     Buying Power
                  </h3>
                  <h3 className="font-semibold text-[26px] text-[#2d405a] buying-power-value">
                     {formatToMoney(Math.trunc(buyingPower))}
                  </h3>
               </div>
            </div>
         )}
      </>
   )
}
