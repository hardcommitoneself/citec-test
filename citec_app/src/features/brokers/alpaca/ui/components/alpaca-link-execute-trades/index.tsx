// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import {
   getAllStatus,
   getTickers,
   processBuy,
   removeCash,
} from 'features/brokers/alpaca/ui/components/alpaca-link-execute-trades/execute-trades-utils'
import { TradeAlpaca } from 'features/brokers/alpaca/utils/trade-alpaca'
import { cancelPosition } from 'features/brokers/alpaca/utils/cancel-position-alpaca'
import { useAppSelector } from 'store/hooks'
import { Button } from '@mui/material'
import { executeOperations } from 'features/brokers/alpaca/ui/components/alpaca-link-execute-trades/execute-trades-utils'

import Swal from 'sweetalert2'
import { useAppNavigate } from 'features/utils/hooks'
import { validateAssetsExchange } from 'features/brokers/utils/validate-assets-exchange'
import { messageHighMape } from 'features/brokers/utils/message-high-mape'


export type linkExecuteTradesProps = {
   data: any
   mape: number
}

export const LinkExecuteTrades = ({ data, mape }: linkExecuteTradesProps) => {
   const visibility = process.env.REACT_APP_ALPACA_VISIBILITY

   const navigate = useAppNavigate()

   const brokers = useAppSelector((state) => state.brokers) 

   const executeTrades = (data: any) => {

      const tickers = getTickers(removeCash(data))
      const isValid = validateAssetsExchange(brokers.accountTypeSelect, tickers)

      isValid.then((valid) => {
         if (valid) {
            Swal.fire({
               title: 'Execute trades',
               text: 'Are you sure you want to execute these trades?',
               icon: 'warning',
               showCancelButton: false,
               confirmButtonText: 'Yes, Execute it!',
               cancelButtonText: '<span style="color:#79828D">No, cancel!</span>',
               showLoaderOnConfirm: true,
               confirmButtonColor: '#9C2B85',
               reverseButtons: true,
            }).then((result) => {
               if (result.isConfirmed) {
                  Swal.fire({
                     title: 'Execute trades',
                     html: `Processing the orders:
                     <ul>                  
                        <li><b>Sell</b>: - / - trades</li>
                        <li><b>Buy</b>: - / - trades</li>
                     </ul>`,
                     allowOutsideClick: false,
                     didOpen: async () => {
                        Swal.showLoading()
                        const { sell, buy } = await executeOperations(data)
                        const totalBuy = buy.length
                        const totalSell = sell.length
                        let valBuy = 0
                        let valSell = 0
                        const ordersIds: string[] = []
      
                        sell.map(async (s) => {
                           const resultSell = await cancelPosition(
                              brokers.accountTypeSelect,
                              s.symbol
                           )
      
                           ordersIds.push(
                              resultSell !== undefined ? resultSell.data.id : 'error'
                           )
      
                           valSell++
                           Swal.update({
                              html: `Processing the orders:
                                    <ul>                                 
                                       <li><b>Sell</b>: ${valSell} / ${totalSell} trades</li>
                                       <li><b>Buy</b>: ${valBuy} / ${totalBuy} trades</li>
                                    </ul>`,
                           })
                        })
      
                        setTimeout(() => {
                           const statusOrders = getAllStatus(
                              brokers.accountTypeSelect,
                              ordersIds
                           )
      
                           const proceedToBuy = processBuy(statusOrders)
                           if (proceedToBuy) {
                              buy.map(async (b) => {
                                 await TradeAlpaca(brokers.accountTypeSelect, b)
                                 valBuy++
                                 Swal.update({
                                    html: `Processing the orders:
                                          <ul>                                    
                                             <li><b>Sell</b>: ${valSell} / ${totalSell} trades</li>
                                             <li><b>Buy</b>: ${valBuy} / ${totalBuy} trades</li>
                                          </ul>`,
                                 })
                              })
                           }
                        }, 1000)
      
                        Swal.hideLoading()
                        Swal.update({
                           showCancelButton: true,
                           showConfirmButton: false,
                           cancelButtonText: '<span style="color:#79828D">Close</span>',
                           cancelButtonColor: '#E8EFF6',
                        })
                        navigate('/')
                     },
                  })
               }
            })
         } else {
            Swal.fire({
               title: "Trades can't be executed",
               text: "Trades can't be executed because at least one of the Alpaca market assets does not match",
               icon: 'warning',
               showCancelButton: false,
               confirmButtonText: 'Ok!',
               confirmButtonColor: '#9C2B85',
               allowOutsideClick: true,
               backdrop: true
            })
         }
      })
      
   }

   if (mape > 5) {
      messageHighMape(mape)
   }

   return (
      <>
         {visibility !== 'undefined' && visibility == 'true' && (
            <div className='inline mr-3.5'>
               <Button
                  variant="actionButtonCitec"
                  className= "h-[35px] min-w-0"
                  onClick={() => executeTrades(data)}
                  data-testid="test-execute-trades"
               >
                  Execute trades
               </Button>
            </div>
            
         )}
      </>
   )
}
