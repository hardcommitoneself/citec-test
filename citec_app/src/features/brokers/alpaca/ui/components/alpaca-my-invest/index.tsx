// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { useState, useEffect } from 'react'
import { Button } from '@mui/material'
import { useAppSelector } from 'store/hooks'
import { InvestTable } from 'features/brokers/utils/table/invest-table'
import { columns } from 'features/brokers/utils/table/my-invest'
import {
   getStockForValidation,
   buildDataInvest,
} from 'features/brokers/alpaca/utils/positions-alpaca-utils'
import { useDispatch } from 'react-redux'
import { setPortfolio } from 'store/slices/portfolio'
import { setWorkflow } from 'store/slices/workflow'
import { PositionAlpaca } from 'features/brokers/alpaca/utils/position-alpaca'
import { getAssetType } from 'features/brokers/utils/get-asset-type'
import { formatInvestmentData } from 'features/brokers/alpaca/utils/format-investment-data'
import XLSX from 'xlsx'
import { useAppNavigate } from 'features/utils/hooks'
import { useSelectUniverse } from 'features/utils/hooks'


export interface MyInvestments {
   asset_class: string
   quantity: number
   price: number
   investment: number
   day_gain_loss: number[]
   total_gain_loss: number[]
   url_logo?: string
   subRows: any
}

export let loadPortfolioOnRedux: any
export let handleOnExportInvestments: any

export const AlpacaMyInvest = () => {
   const visibility = process.env.REACT_APP_ALPACA_VISIBILITY

   const brokers = useAppSelector((state) => state.brokers)

   const [dataTable, setDataTable] = useState<MyInvestments[]>([])

   const navigate = useAppNavigate()
   const dispatch = useDispatch()

   const { handleSelectBenckmark } = useSelectUniverse()

   const dataInvest = async () => {
      const positions = await PositionAlpaca(brokers.accountTypeSelect)

      const stocks = getStockForValidation(positions)
      const assetType = await getAssetType(stocks)
      const response = buildDataInvest(positions, assetType)
      setDataTable(response)
   }

   useEffect(() => {
      dataInvest()
   }, [brokers.accountTypeSelect])

   loadPortfolioOnRedux = (row: any, flow: string) => {

      if (flow === 'rebalance') {
         dispatch(setWorkflow('rebalance'))
      } else {
         dispatch(setWorkflow('load'))
      }

      const ticker = []
      const outstanding_balance = []
      const unrealized_capital_gains = []
      for (let i = 0; i < row.length; i++) {
         ticker.push(row[i].values.asset_class)
         outstanding_balance.push(Number(row[i].values.investment))
         unrealized_capital_gains.push(Number(row[i].values.total_gain_loss[0]))
      }
      dispatch(
         setPortfolio({
            ticker,
            outstanding_balance,
            unrealized_capital_gains,
         })
      )

      handleSelectBenckmark('SPY')

      if (flow === 'load') {
         navigate('/portfolio-loader')
      } else if (flow === 'analysis') {
         navigate('/portfolio-analysis')
      } else if (flow === 'rebalance') {
         navigate('/investment')
      }
   }

   handleOnExportInvestments = (row: any) => {
      try {
         const data_investment = []
         data_investment.push(formatInvestmentData(row.values))

         for (const i of row.subRows) {
            data_investment.push(formatInvestmentData(i.values))
         }

         const wordBook = XLSX.utils.book_new()
         const wordSheet = XLSX.utils.json_to_sheet(data_investment)

         XLSX.utils.book_append_sheet(wordBook, wordSheet)
         XLSX.writeFile(wordBook, 'my_investments.xlsx')

         return 1
      } catch (error) {
         return error
      }
   }

   return (
      <>
         {visibility == 'true' && (
            <div className="mx-10 mt-8 bg-white p-4 rounded-2xl relative balance">
               <div className="flex buying-power">
                  <h3 className="font-semibold text-[26px] text-[#2d405a]">
                     My Investments
                  </h3>
               </div>
               <InvestTable columns={columns} data={dataTable} messageNoData="" />
            </div>
         )}
      </>
   )
}
