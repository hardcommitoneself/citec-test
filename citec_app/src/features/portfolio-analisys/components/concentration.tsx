// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import ReactApexChart from 'react-apexcharts'

export const Concentration = (data: {
   [x: string]: { [x: string]: string | any[] }
}) => {
   const series: Array<any> = [
      {
         name: 'weight',
         data: data['concentration']['concentration'],
      },
   ]

   const stylesFonts = {
      color: '#6C8394',
      fontSize: '14px',
      fontFamily: 'Poppins',
      fontWeight: 'normal',
   }

   const options: ApexCharts.ApexOptions = {
      chart: {
         height: 350,
         type: 'line' as const,
         zoom: {
            enabled: false,
         },
      },
      dataLabels: {
         enabled: false,
      },
      tooltip: {
         //https://apexcharts.com/docs/options/tooltip/
         enabled: true,
         x: {
            show: false,
         },
         marker: {
            show: false,
         },
      },
      stroke: {
         //https://apexcharts.com/docs/options/stroke/
         curve: 'straight' as const,
         colors: ['#20BBB7'],
      },
      markers: {
         //https://apexcharts.com/docs/options/markers/
         colors: ['#20BBB7'],
      },
      grid: {
         row: {
            colors: ['#f3f3f3', 'transparent'], // takes an array which will be repeated on columns
            opacity: 0.5,
         },
      },
      xaxis: {
         // apexcharts.com/docs/options/xaxis/
         categories: [],
         tickAmount: 10,
         title: {
            text: 'Number of stocks',
            offsetX: 0,
            offsetY: 5,
            style: {
               ...stylesFonts,
            },
         },
         labels: {
            show: true,
            rotate: 0,
            style: {
               ...stylesFonts,
            },
         },
      },
      yaxis: {
         // apexcharts.com/docs/options/yaxis/
         floating: false,
         decimalsInFloat: 0,
         title: {
            text: '',
            offsetX: 0,
            offsetY: 0,
         },
         labels: {
            style: {
               ...stylesFonts,
            },
         },
      },
   }

   return (
      <div>
         <ReactApexChart
            options={options}
            series={series}
            type="line"
            height={450}
         />
      </div>
   )
}
