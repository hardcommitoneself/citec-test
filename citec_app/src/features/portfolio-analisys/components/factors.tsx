// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import ReactApexChart from 'react-apexcharts'
import converterArrays from 'features/utils/converterArrays'
import { numberToFixed } from 'features/utils/to-fixed-number'

export const Factors = ({ factors }: any) => {
   const dataSeriesMultipliedBy100: { x: any; y: number }[] = []

   const dataSeries = converterArrays({
      x: factors['factor'],
      y: factors['percentage'],
   })
   dataSeries.map(function (row) {
      dataSeriesMultipliedBy100.push({
         x: row.x,
         y: numberToFixed(row.y * 100, 0),
      })
   })

   const series = [
      {
         data: dataSeriesMultipliedBy100,
      },
   ]

   const options = {
      chart: {
         height: 350,
         type: 'bar' as const,
      },
      colors: [
         function ({ dataPointIndex }: any) {
            if (dataPointIndex == 0) {
               return '#6576FF'
            } else if (dataPointIndex == 1) {
               return '#FFA6A6'
            } else if (dataPointIndex == 2) {
               return '#56CCF2'
            } else if (dataPointIndex == 3) {
               return '#0BACA9'
            } else if (dataPointIndex == 4) {
               return '#FF9E37'
            } else if (dataPointIndex == 5) {
               return '#8E547F'
            } else {
               return '#83CA76'
            }
         },
      ],
      plotOptions: {
         bar: {
            horizontal: false,
            columnWidth: '55%',
            borderRadius: 10,
            endingShape: 'rounded' as const,
         },
      },
      dataLabels: {
         enabled: false,
      },
      stroke: {
         show: true,
         width: 2,
         colors: ['transparent'],
      },
      grid: {
         //https://apexcharts.com/docs/options/grid/
         show: false,
      },
      yaxis: {
         decimalsInFloat: 0,
         labels: {
            style: {
               colors: [
                  '#79828d',
                  '#79828d',
                  '#79828d',
                  '#79828d',
                  '#79828d',
                  '#79828d',
                  '#79828d',
               ],
               fontSize: '14px',
               fontFamily: 'Poppins',
               cssClass: 'apexcharts-font-citec',
            },
         },
      },
      xaxis: {
         labels: {
            rotateAlways: true,
            rotate: -90,
            style: {
               colors: [
                  '#79828d',
                  '#79828d',
                  '#79828d',
                  '#79828d',
                  '#79828d',
                  '#79828d',
                  '#79828d',
               ],
               fontFamily: 'Poppins',
               cssClass: 'apexcharts-font-citec',
               fontSize: '14px',
            },
         },
         axisTicks: {
            show: true,
         },
      },
      tooltip: {
         //https://apexcharts.com/docs/options/tooltip/
         enabled: true,
         x: {
            show: false,
         },
         y: {
            title: {
               formatter: () => '',
            },
         },
         marker: {
            show: true,
         },
      },
   }

   return (
      <ReactApexChart options={options} series={series} type="bar" height={450} />
   )
}
