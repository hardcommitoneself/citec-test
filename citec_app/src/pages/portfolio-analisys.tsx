// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { useDispatch } from 'react-redux'
import { useEffect, useState } from 'react'
import Button from '@mui/material/Button'

import { BacktestRes } from 'features/utils/interfaces/backtest-api-resp'

import { Loading } from 'features/ui'
import { PageContainer } from 'features/ui/page-container'
import { PortfolioAnalisysResp } from 'features/utils/interfaces/portfolio-analysis-resp'
import { PositionsResp } from 'features/utils/interfaces/positions-resp'
import { setPortfolioAnalysis, setPortfolioPositions } from 'store/slices/portfolio'
import { useAppNavigate, useFetchApi } from 'features/utils/hooks'
import { useAppSelector } from 'store/hooks'
import {
   RiskreturnTab,
   Backtest,
   OverviewTab,
   Positions,
} from 'features/portfolio-analisys/tabs'
import { savePortfolio } from 'features/utils/save-portfolio'

export const PortfolioAnalysis = () => {
   const navigate = useAppNavigate()

   const { ticker, outstanding_balance, unrealized_capital_gains } = useAppSelector(
      (state) => state.portfolio
   )

   const outstanding_balance_total = outstanding_balance?.reduce(
      (accumulator, curr) => accumulator + curr
   )

   const benchmark = useAppSelector((state) => state.rebalance.benchmark)

   const { resp, isLoading, error } = useFetchApi<PortfolioAnalisysResp>({
      endpoint: '/ai/portfolio/analysis',
      method: 'POST',
      params: {
         hist_w: 5000,
      },
      payload: {
         portfolio: 'myportfolio',
         benchmark,
         ticker,
         outstanding_balance,
         unrealized_capital_gains,
      },
   })
   const { resp: backtest } = useFetchApi<BacktestRes>({
      endpoint: '/ai/portfolio/backtest',
      method: 'POST',
      params: {
         hist_w: 365,
      },
      payload: {
         portfolio: 'myportfolio',
         benchmark,
         ticker,
         outstanding_balance,
         unrealized_capital_gains,
      },
   })

   const { resp: positions } = useFetchApi<PositionsResp>({
      endpoint: '/ai/portfolio/positions',
      method: 'POST',
      payload: {
         portfolio: 'myportfolio',
         benchmark,
         ticker,
         outstanding_balance,
         unrealized_capital_gains,
      },
   })

   const handleSavePortfolio = () => {
      savePortfolio({
         navigate,
         benchmark,
         outstanding_balance,
         ticker,
         unrealized_capital_gains,
      })
   }

   const [active, setActive] = useState('Positions')
   const dispatch = useDispatch()

   const views = ['Positions', 'Overview', 'Risk', 'Backtest']

   useEffect(() => {
      if (resp) {
         dispatch(setPortfolioAnalysis({ ...resp }))
      }
      if (positions) {
         const newArr1 = positions.map((v) => ({
            ...v,
            weight: (
               (v.outstanding_balance * 100) /
               outstanding_balance_total
            ).toFixed(2),
         }))

         dispatch(setPortfolioPositions(newArr1))
      }
   }, [resp])

   const enabledViews = ['Positions', 'Risk', 'Overview', 'Backtest']

   if (isLoading) return <Loading />

   return (
      <PageContainer error={error}>
         <div className="InvestmentBody summary dashboard_body_portfolio_analysis">
            <div className="inner_dashboard_cri">
               <div className="options_summary_buttons">
                  {views.map((view) => (
                     <button
                        key={view}
                        className={active === view ? 'active-btn' : 'null'}
                        disabled={
                           //disabled is temporal, when the functionality is ready must remove this line
                           !enabledViews.includes(view)
                        }
                        onClick={() => {
                           setActive(view)
                        }}
                        data-testid={`btn-${view.toLowerCase}`}
                     >
                        {view}
                     </button>
                  ))}
               </div>
               {resp && backtest && (
                  <>
                     {active === 'Positions' && <Positions backtest={backtest} />}
                     {active === 'Backtest' && <Backtest backtest={backtest} />}
                     {active === 'Risk' && <RiskreturnTab />}
                     {active === 'Overview' && <OverviewTab backtest={backtest} />}
                  </>
               )}
               <div className="buttons_wrapper">
                  <Button
                     className="w-28 h-12"
                     color="secondary"
                     onClick={() => navigate('goBack')}
                  >
                     <span className="text-base font-normal">Back</span>
                  </Button>

                  <Button
                     className="w-28 h-12"
                     variant="save"
                     onClick={handleSavePortfolio}
                  >
                     <span>Save</span>
                  </Button>
               </div>
            </div>
         </div>
      </PageContainer>
   )
}
