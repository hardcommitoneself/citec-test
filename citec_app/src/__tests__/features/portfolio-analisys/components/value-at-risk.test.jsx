// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { render } from '@testing-library/react'
import { ValueAtRisk } from 'features/portfolio-analisys/components'

class ResizeObserver {
   observe() {}
   unobserve() {}
}
jest.mock('react-apexcharts', () => {
   return {
      __esModule: true,
      default: () => {
         return <div />
      },
   }
})

const valueatrisk = {
   quantile: [0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100],
   value: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 3.0, 10.0],
}

describe('Value at Risk component', () => {
   test('Snapshot', async () => {
      window.ResizeObserver = ResizeObserver

      const { container } = render(<ValueAtRisk valueatrisk={valueatrisk} />)

      expect(container).toMatchSnapshot()
   })
})
