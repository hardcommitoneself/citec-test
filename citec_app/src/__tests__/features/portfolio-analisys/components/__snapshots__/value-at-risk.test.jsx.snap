// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Value at Risk component Snapshot 1`] = `
<div>
  <div />
</div>
`;
